### ISSUE TITLE
Consider using more Singleton/Bind at Service Container to handle classes instances

### ISSUE THREAD

**Issue by henriqueramos** [pkp/pkp-lib#7282]
Back on pkp/pkp-lib#7125, I had a problem with `IsoCodesFactory` showing countries names list on Portuguese version, even the OJS installation had `en_US` and `fr_CA` locales as only languages. Thus my local test scenarios aren't being able to pass.

This was due to the way `IsoCodesFactory` implement his translation driver, getting the current language value directly from the `LANGUAGE` environment variable, and not from the `LC_ALL`, as we did on [PKPLocale::initialize](https://github.com/pkp/pkp-lib/blob/main/classes/i18n/PKPLocale.inc.php#L339).

I've fix this issue, by creating a [Singleton](https://refactoring.guru/design-patterns/singleton) for the `IsoCodesFactory`, setting the locale driver same as the PKP locale, and calling this instance across the system.

i.e

`PKPContainer::registerBaseBindings`
```php
$this->singleton(IsoCodesFactory::class, function () {
    $driver = new GettextExtensionDriver();
    $driver->setLocale(PKPLocale::getLocale());
    return new IsoCodesFactory(null, $driver);
});
```

```php
$isoCodes = app(IsoCodesFactory::class);

dd((array) $isoCodes->getCountries());
```

But we should rely more on Singletons and Binds at the [Service Container](https://github.com/pkp/pkp-lib/blob/main/classes/core/PKPContainer.inc.php), empowering us to control our classes instances better.

**Comment by NateWr** [comment]
I agree. It took me a long time to get my head around the Service Container, but once I did I realized that it is going to play a huge role going forward in helping us manage dependencies across all three applications. It also offers us a great way to expose some of the core pieces of the software to extension by plugins.

We should aim to integrate many of our core systems (eg - `PKPLocale`, `Application`, `TemplateManager`, `Router`, `Dispatcher`) into the service container, as well as all of the existing service classes that do not fit the Repo/EntityDAO structure (`FileService`, `PKPSchemaService`, `StatsService`, `StatsEditorialService`). The latter conversion is filed at https://github.com/pkp/pkp-lib/issues/7131.

@asmecher I know there's already a lot to consider at the moment, but if you get a chance to explore the Service Container system -- and especially the combination of its dependency injection and binding features -- I think there's a lot of opportunity here. Off the top of my head, it would be an appropriate place to provide support for things like remote file support, third-party mail services, caching mechanisms, etc. The nice thing about the container approach is that a plugin can inject one class instead of another, so overriding behaviour at a low-level is often as simple as extending a class, overriding methods, and binding that class to calls for another.

**Comment by asmecher** [comment]
@NateWr, yep, Henrique and I were going over this on Monday, which is what led to this filing. I'm still in the getting-my-head-around it phase but am very interested in pushing our use further.

**Comment by NateWr** [comment]
So far, I've been putting our core dependencies into an `AppServiceProvider`. You can see here how I've used this to override `PKPRequest` with `Request`.

pkp-lib: https://github.com/pkp/pkp-lib/blob/main/classes/core/AppServiceProvider.inc.php#L45-L47
ojs: https://github.com/pkp/ojs/blob/main/classes/core/AppServiceProvider.inc.php#L30

**Comment by jonasraoni** [comment]
Talking about the container, we can already get rid of the `Pimple` dependency =]
As a side note, I'm using the container to serve the locale for the #6328:

```php
public function register(): void
{
    $this->app->singleton(LocaleInterface::class, function () {
        return $this->app->make(Locale::class);
    });
    // Replaces the default Laravel translator
    $this->app->alias(LocaleInterface::class, 'translator');
    $this->app->alias(LocaleInterface::class, Translator::class);
}
```

To keep the ability of using the `Locale` statically (e.g. `Locale::getLocale()`), I've created a `Facade` for it:
```php
class Locale extends \Illuminate\Support\Facades\Facade
{
    protected static function getFacadeAccessor(): string
    {
        return LocaleInterface::class;
    }
}
```

About testing, the `Facade` provides mocking methods  and it's even possible to replace the underlying `Locale` instance on the fly with `Locale::swap(new MockedLocale())`.

**Comment by NateWr** [comment]
That's cool, @jonasraoni! When extending `Illuminate\Support\Facades\Facade` and using `getFacadeAccessor`, is your code editor able to pick up the available methods for autocomplete? Like if you type `Locale::get` does it show you that `getLocale()` is a method on the class?

I think that was one of the hurdles I ran into when making the `Repo` facade, but maybe I was doing something wrong.

**Comment by jonasraoni** [comment]
Yeah! That's the stinky part of the `Facade`, which I was going to discuss after opening a PR. As it just has a `__callStatic`, other limitations include: accessing properties, constants, passing it as an argument, etc.

FYI [Laravel is using a docblock](https://github.com/illuminate/support/blob/4ee2d8fb589731864545e93aa6a4cb5ec519377a/Facades/Bus.php#L10-L26) to help the code editor display the available methods.

---

I think the `Facade` is handy, but I would personally opt-out from it and follow a more standard/typed approach, so I think it's good opportunity to discuss a pattern here :)

I think the most straightforward way is to have accessors, like what you did in the `PKP\facades\Repo`. For the locale, I could add a `Locale::get()` to avoid using the untyped `app(Locale::class)`.

**Comment by NateWr** [comment]
That all sounds good. If we are going to go this route, I wonder if it makes sense to step back and think about when we want to use facades and when we should rely on dependency injection. In general, I think dependency injection is preferred. 

But when it comes to building extensibility into the system, facades are really useful. For example, I'd be tempted to thing that a `Locale` class shouldn't be need outside of a few core classes like the request and the application. But I suspect that there will be cases where a plugin or event of some kind wants access to it. Is it better to have a wrapping class, like `Application`, that provides access to these? Or to use a facade of some kind to get a singleton?

**Comment by jonasraoni** [comment]
Hmm, I support using shortcuts (in my case, it will be `Locale::getInstance()`) for now, and "wire" the classes with dependency injection as we can.

About the `Laravel's Facade`, which I've used mostly to keep things similar, I think it's better to discard 🤔
- The only use case I see is to enforce all code base is using the active instance (e.g. if N classes receive a `Request` instance by DI, then we replace the request in the container, they will keep using the old one).
- I prefer to use such dynamic features of the language for dynamic tasks (e.g. navigating through a JSON/XML).

**Comment by NateWr** [comment]
The only downside to using something like `Locale::getInstance()` is that the class can not be overwritten. Making use of Laravel's service containers we can do something like `App::make(Locale::class)` and plugins can inject their own extended class at run-time...

**Comment by jonasraoni** [comment]
Hmm, I think this way is ok:

```php
public static function getInstance(): LocaleInterface
{
    return App::make(LocaleInterface::class);
}
```

The class is just serving as a shortcut to get a type-hinted instance, and a plugin might still replace the `LocaleInterface` in the container by an extended version.

**Comment by asmecher** [comment]
Merged the PRs above, @henriqueramos! Does this need to be ported to OMP or OPS?
